picasso/src/main/java/com/squareup/picasso3/MatrixTransformation.java

```diff
@@ -0,0 +1,229 @@
+package com.squareup.picasso3;
+
+import android.graphics.Bitmap;
+import android.graphics.Matrix;
+import android.view.Gravity;
+
+import static android.support.media.ExifInterface.ORIENTATION_FLIP_HORIZONTAL;
+import static android.support.media.ExifInterface.ORIENTATION_FLIP_VERTICAL;
+import static android.support.media.ExifInterface.ORIENTATION_ROTATE_180;
+import static android.support.media.ExifInterface.ORIENTATION_ROTATE_270;
+import static android.support.media.ExifInterface.ORIENTATION_ROTATE_90;
+import static android.support.media.ExifInterface.ORIENTATION_TRANSPOSE;
+import static android.support.media.ExifInterface.ORIENTATION_TRANSVERSE;
+
+public class MatrixTransformation implements Transformation {
+  private final Request data;
+
+  public MatrixTransformation(Request data) {
+    this.data = data;
+  }
+
+  @Override public RequestHandler.Result transform(RequestHandler.Result source) {
+    if (!source.hasBitmap()) {
+      return source;
+    }
+
+    Bitmap bitmap = transformResult(data, source.getBitmap(), source.getExifOrientation());
+
+    return new RequestHandler.Result(bitmap, source.getLoadedFrom(), source.getExifOrientation());
+  }
+
+  @Override public String key() {
+    return null;
```

Comment:
This works? We need to throw if someone does this. Can be a follow-up.

If we're going to make this behavior into a transformation then we should move the code which adds the data to they key into this class and return a proper key. Can also be a follow-up.

Reply:
Making non-null string value for now.  Can you elaborate more on "then we should move the code which adds the data to they key into this class and return a proper key"?

Reply:
Keys are supposed to contain the data which makes them unique. If you have a rotate transformation the key needs to contain the rotation information. Otherwise the cache will return a 180 rotated image for someone requesting 90.

In this case I was saying we should update they key to contain the necessary information that makes the transformation unique.

Reply:
got it.  updated to "matrixTransformation()".